vod/hls/m3u8_builder.c

```diff
@@ -565,78 +573,89 @@ m3u8_builder_build_master_playlist(
 	}
 
 	// write the header
-	p = vod_copy(result->data, m3u8_header, sizeof(m3u8_header) - 1);
 
+	p = vod_copy(result->data, m3u8_header, sizeof(m3u8_header) - 1);
 	// write the streams
 	for (cur_sequence = media_set->sequences; cur_sequence < media_set->sequences_end; cur_sequence++)
 	{
-		cur_sequence_tracks = cur_sequence->filtered_clips[0].longest_track;
-
-		// write the track information
-		if (cur_sequence_tracks[MEDIA_TYPE_VIDEO] != NULL)
-		{
-			track = cur_sequence_tracks[MEDIA_TYPE_VIDEO];
-			video = &track->media_info;
-			bitrate = video->bitrate;
-			if (cur_sequence_tracks[MEDIA_TYPE_AUDIO] != NULL)
+		main_media_type = cur_sequence->track_count[MEDIA_TYPE_VIDEO] != 0 ? MEDIA_TYPE_VIDEO : MEDIA_TYPE_AUDIO;
+             
+             
+		cur_sequence_tracks[MEDIA_TYPE_AUDIO] = cur_sequence->filtered_clips[0].longest_track[MEDIA_TYPE_AUDIO];
+		for (cur_track = cur_sequence->filtered_clips[0].first_track; cur_track < cur_sequence->filtered_clips[0].last_track; cur_track++)
+		{		
+			cur_sequence_tracks[MEDIA_TYPE_VIDEO] = NULL;
+			//cur_sequence_tracks[MEDIA_TYPE_AUDIO] = NULL;	
+			if (cur_track->media_info.media_type != main_media_type)
 			{
-				audio = &cur_sequence_tracks[MEDIA_TYPE_AUDIO]->media_info;
-				bitrate += audio->bitrate;
+				continue;
 			}
-			p = vod_sprintf(p, m3u8_stream_inf_video,
-				bitrate,
-				(uint32_t)video->u.video.width,
-				(uint32_t)video->u.video.height,
-				&video->codec_name);
-			if (cur_sequence_tracks[MEDIA_TYPE_AUDIO] != NULL)
+			cur_sequence_tracks[cur_track->media_info.media_type] = cur_track;
+			// write the track information
+			if (cur_sequence_tracks[MEDIA_TYPE_VIDEO] != NULL)
 			{
-				*p++ = ',';
-				p = vod_copy(p, audio->codec_name.data, audio->codec_name.len);
+				video = &cur_track->media_info;
+				bitrate = video->bitrate;
+				if (cur_sequence_tracks[MEDIA_TYPE_AUDIO] != NULL)
+				{
+					audio = &cur_sequence_tracks[MEDIA_TYPE_AUDIO]->media_info;
+					bitrate += audio->bitrate;
+				}
+				p = vod_sprintf(p, m3u8_stream_inf_video,
+						bitrate,
+						(uint32_t)video->u.video.width,
+						(uint32_t)video->u.video.height,
+						&video->codec_name);
+				if (cur_sequence_tracks[MEDIA_TYPE_AUDIO] != NULL)
+				{
+					*p++ = ',';
+					p = vod_copy(p, audio->codec_name.data, audio->codec_name.len);
+				}
+			}
+			else
+			{
+				cur_track = cur_sequence_tracks[MEDIA_TYPE_AUDIO];
```

Comment:
you are updating the loop variable inside the loop, this a bad idea and in this case unneeded, just use cur_track as is